In [6]:
import requests
import json
import re

# Function to extract a 6-digit PIN code from the address string
def extract_pincode(address):
    # Define a regex pattern to match 6-digit PIN codes
    pattern = r"\b\d{6}\b"
    # Use regex to find all 6-digit numbers in the address string
    matches = re.findall(pattern, address)
    # Return the last match as the PIN code
    return matches[-1] if matches else ""

# Function to validate a PIN code by checking if it corresponds to the given address
def validate_pincode(pincode, address):
    if len(pincode) == 6:
        # Create the URL for the postalpincode.in API
        api_url = f"https://api.postalpincode.in/pincode/{pincode}"
        # Send a GET request to fetch information for the PIN code
        response = requests.get(api_url)

        if response.status_code == 200:
            data = response.json()
            # Check if the API response indicates success
            if data and data[0]['Status'] == 'Success':
                # Iterate through post offices associated with the PIN code
                for post_office in data[0]['PostOffice']:
                    # Check if the post office name is found in the address (case-insensitive)
                    if post_office['Name'].lower() in address.lower():
                        return True
    return False

# Main function to take user input and perform address validation
def main():
    address = input("Enter your address:\n")
    pincode = extract_pincode(address)

    if validate_pincode(pincode, address):
        print("Address is correct")
    else:
        print("Address is not correct")

if __name__ == "__main__":
    main()


Enter your address:
Colony, Bengaluru, Karnataka 560095
Address is not correct


Correct addresses:

*   2nd Phase, 374/B, 80 Feet Rd, Mysore Bank Colony, Banashankari 3rd Stage, Srinivasa Nagar, Bengaluru, Karnataka 560050
*   80 Feet Rd, State Bank Colony, Banashankari 3rd Stage, Srinivasa Nagar, Bangalore. 560050
*   21-19/1, New gunj, Kosgi, 509339

wrong adresses:

*   2nd Phase, 374/B, 80 Feet Rd, Mysore Bank Colony, Banashankari 3rd Stage, Srinivasa Nagar, Bengaluru, Karnataka 560095
*   Colony, Bengaluru, Karnataka 560095
*   Address with no pin
*   empty string









